### PR TITLE
Reintroduce aspect ratio logic for tiling containers

### LINKED ISSUE #3795: Content of workspaces (screen background) visible on others
## I'm submitting a…

<pre>
[x] Bug
[ ] Feature Request
[ ] Documentation Request
[ ] Other (Please describe in detail)
</pre>

## Current Behavior
i3  4.17.1: Background of one workspace is visible, when switching to a different workspace.

## Expected Behavior
When switching from one workspace, the contents that was seen on the former workspace is not visible anymore.

## Reproduction Instructions
Example reproduction:
One workspace with alsamixer that fills the screen (one tile only).
Then switch to a different workspace (I had more than one tile, but I think this does not matter here). If the used tile is not completely used by the program on the newly encountered (switched to) workspace, the contents of the former worksapce frames the new workspace.

## Environment
<pre>
Arch Linux.
Package that works is 4.16.1-1.
Package with the problem is 4.17.1-1.
</pre>

After finding the problem I downgraded to the older version to be sure, that no other package-installations have caused the problem.
I found out that with downgrading to 4.16.1-1 the problem was gone.
So the problem is a i3-problem.

Output of `i3 --moreversion 2>&-`:
<pre>
i3 version: 
Binary i3 version:  4.16.1 (2019-01-27) © 2009 Michael Stapelberg and contributors
Running i3 version: 4.16.1 (2019-01-27) (pid 28339)bort…)
Loaded i3 config: /home/oliver/.i3/config (Last modified: Sa 02 Mär 2019 23:50:22 CET, 17187316 seconds ago)

The i3 binary you just called: /usr/bin/i3
The i3 binary you are running: i3
</pre>

See discussion on i3-mailinglist:
https://www.freelists.org/archive/i3-discuss/09-2019

Screenshots:

a) with the problem:
http://s000.tinyupload.com/?file_id=53830511177154834717

b) after downgrading, expected behaviour:
http://s000.tinyupload.com/index.php?file_id=16308606597116519931

### PR DESCRIPTION
**Pull request by orestisfl**
Fixes #3795

### CODE REVIEW

**Comment by xzfc** [comment]
After this commit, mpv aspect ratio hints are honored as well (unlike 4.16).

Ignoring mpv aspect ratio hints was a useful feature: mpv could display OSD outside of the movie frame. And this was consistent with i3 behavior regarding "resize increment" hints for tiled windows.

---

I've checked it using `slop -b 20` across commits:

| commit | mpv | mplayer |
| ---    | --- | ---     |
| 29f2510fa9484a5fa235de9a1c5c937292151888^ | inored  | honored, no bug |
| 29f2510fa9484a5fa235de9a1c5c937292151888  | ignored | ignored, bug |
| 6b8b3dcc8927179d45df99f61958f881f12e1ac7^ | ignored | ignored, bug |
| 6b8b3dcc8927179d45df99f61958f881f12e1ac7  | honored | honored¹, no bug |

1) Aspect ratio hints are ignored right after mplayer startup but honored after workspace switching.

**Comment by orestisfl** [comment]
Thank you very much! I'll investigate.

**Comment by orestisfl** [comment]
I think that the problem with mpv is that before https://github.com/i3/i3/pull/3432/commits/29f2510fa9484a5fa235de9a1c5c937292151888 i3 did not save the aspect ratio normal hints in `manage_window` so mpv had `window->aspect_ratio == 0.0`.

I don't think there is any reasonable way to correct that. I much prefer to not introduce problems to the more popular mpv player in order to fix a bug in mplayer.


@Airblader any other ideas?

**Comment by Airblader** [comment]
In my understanding the bug lies with mplayer, right? As such I wouldn't try to work around its issues in i3, particularly if it causes issues with other clients (more popular or not).

**Comment by orestisfl** [comment]
In this case, I don't know which is the "correct" behavior. i3 exposes a bug in mplayer but it would be possible for i3 to misbehave according to the standard.

According to the comment by Michael:
https://github.com/i3/i3/blob/f397698d43e3dd4f4625e1fed1b1d0470d0acdf6/src/render.c#L67-L75
the spec is not clear about fullscreen windows (and certainly not about "tiling" windows) but in this specific case I feel that we should ignore the aspect ratio hints (as in 4.17) like we do for the normal hints.

**Comment by daviddavo** [comment]
Could this be activated with a keyboard shortcut or some kind of action? Having a window locked to an aspect ratio is pretty useful even with tiling containers

**Comment by xsawyerx** [comment]
I'm suffering from the issue this PR is fixing. I was wondering if there are plans to continue the work and merge this fix or a different one?

**Comment by orestisfl** [comment]
I will gladly make the changes if we reach an agreement.

@stapelberg do you have an opinion on this?

**Comment by stapelberg** [comment]
Could you summarize what the question is please? I have read this issue, but it’s not 100% clear to me.

In particular, https://github.com/i3/i3/pull/3810#issuecomment-537787605 says that with your latest commit, both mpv and mplayer do work as expected (isn’t that the desired outcome?), or am I reading this wrong?

---

PS: Hey @xsawyerx! I remember reading your name many times when I was working with Perl. Thanks for all your work, and enjoy i3! :)

**Comment by xsawyerx** [comment]
> Could you summarize what the question is please? I have read this issue, but it’s not 100% clear to me.

I don't know if I'm phrasing it well, but, when using mplayer, the window area, past the video border itself, does not stay black. Instead it reflects other windows backgrounds (like the workspace I moved from, the last image on the screen in full screen, etc.).

> In particular, [#3810 (comment)](https://github.com/i3/i3/pull/3810#issuecomment-537787605) says that with your latest commit, both mpv and mplayer do work as expected (isn’t that the desired outcome?), or am I reading this wrong?

I had missed that. I'm certain I'm not using the latest version, so I'll first upgrade and then check again. Apologies if I added more noise to a resolved issue.

> PS: Hey @xsawyerx! I remember reading your name many times when I was working with Perl. Thanks for all your work, and enjoy i3! :)

I've been a very happy i3 user for a long time now. I think it is I who owes *you* a thank you. :)

**Comment by xsawyerx** [comment]
> > In particular, [#3810 (comment)](https://github.com/i3/i3/pull/3810#issuecomment-537787605) says that with your latest commit, both mpv and mplayer do work as expected (isn’t that the desired outcome?), or am I reading this wrong?

I think this commit is not merged, so I'm not sure upgrading from 4.17.1 (current stable Ubuntu) to 4.18.1 (official stable) will fix the issue. Maybe I'm missing something.

**Comment by stapelberg** [comment]
> I think this commit is not merged, so I'm not sure upgrading from 4.17.1 (current stable Ubuntu) to 4.18.1 (official stable) will fix the issue. Maybe I'm missing something.

Correct, the commit in question is pending in this very PR.

If you want to test, check out this PR and build i3 yourself. You can use `apt build-dep i3-wm` to get all the deps, and `mkdir -p build && cd build && meson .. && ninja` to build i3 :)

**Comment by orestisfl** [comment]
> Could you summarize what the question is please? I have read this issue, but it’s not 100% clear to me.
> 
> In particular, [#3810 (comment)](https://github.com/i3/i3/pull/3810#issuecomment-537787605) says that with your latest commit, both mpv and mplayer do work as expected (isn’t that the desired outcome?), or am I reading this wrong?
> 
> PS: Hey @xsawyerx! I remember reading your name many times when I was working with Perl. Thanks for all your work, and enjoy i3! :)

"Honored" for mpv is not the optimal behavior as explained because subtitles can be placed in the dark areas that appear when the aspect ratios are not honored. I wonder if mpv should report different aspect ratios here because it clearly does not need/want to be that restricted.

**Comment by stapelberg** [comment]
> "Honored" for mpv is not the optimal behavior as explained because subtitles can be placed in the dark areas that appear when the aspect ratios are not honored. I wonder if mpv should report different aspect ratios here because it clearly does not need/want to be that restricted.

Okay, so it’s not optimal because a welcome side-effect of ignoring the aspect ratio hints is no longer triggered. But, it fixes behavior otherwise. Did I get this right?

I’m okay with going ahead with the fix, and requiring people to take up mpv usability improvements with mpv upstream :). Given that mpv is an active open source project, we can fix this the right way instead of having to accept a workaround.

**Comment by orestisfl** [comment]
Ok, I will contact the mpv upstream
